### PR TITLE
[jit] Replace Slot on script::Method with NamedIValue

### PR DESCRIPTION
**Pull request by driazati**
This refactor lets us track the types of initial values added onto a `Method`. The main motivation for this is the change in `module.cpp`.

### CODE REVIEW

File: torch/csrc/jit/script/module.h [hunk 1]
```diff
@@ -461,6 +457,10 @@ struct Module {
     return parameters[name].slot();
   }
 
+  const NamedIValue* get_named_parameter(const std::string& name) const {
+    return parameters.find(name);
```
Comment (eellison) [inline torch/csrc/jit/script/module.h]: Is the intention that this will only ever be called with names that won't fail? Otherwise maybe error here

**Review by eellison — APPROVED** [review]
LGTM. Minor nits / comments


EDIT: looks like tests are breaking so look into that, also please add a test to this ?

File: torch/csrc/jit/script/module.h [hunk 2]
```diff
@@ -52,13 +52,27 @@ struct Module;
 using ModuleLookup = std::function<std::shared_ptr<Module>(
     const std::vector<std::string>&)>;
 
+struct NamedIValue {
```
[the remark this answers is not in the document]
Comment (driazati) [reply torch/csrc/jit/script/module.h]: It's used in `Method`

**Review by zdevito — APPROVED** [review]
<spoke in person>

File: torch/csrc/jit/script/module.h [hunk 3]
```diff
@@ -53,13 +53,35 @@ struct Module;
...
+};
+
 struct Method {
   Method(
       Module* owner,
       std::string name,
       bool optimize,
       std::shared_ptr<Graph> graph,
-      std::vector<Slot> initial_members,
+      std::vector<const NamedIValue*> initial_members,
```
Comment (zdevito) [inline torch/csrc/jit/script/module.h]: NamedIValue's are objects stored continuously in a std::vector. Appending to that vector invalidates the pointers to that vector. So these lists of NamedIValue* are unsafe. If this is true, then this is a major bug and we need to revert.

**Pull request closed without merging**